### PR TITLE
Add short error message-format

### LINKED ISSUE #42653: add --message-format short option
> I copied this from rust-lang/cargo#4165 at @alexcrichton's request

`--message-format` currently has the following options:

```
--message-format FMT         Error format: human, json [default: human]
```

I would like an additional option to be added: `short`. Other possible names:
- `brief`
- `oneline`

I've been programming in rust for the better part of a year now. The compiler has been my friend and mentor through this time, and a big part of that has been the fantastic human readable error messages. Thank you so much for them!

However, I have gotten much more familiar with the language and my common mistakes. Therefore I rarely need the error messages to be so long (sometimes they are longer than 10-15 lines!). Normally I just need the compiler to give the me the file, line/col number and name of the error and I can spot it myself very quickly.

For an experienced programmer, the option of a shorter format would make finding errors much more ergonomic.

The following pipe does *roughly* what I want, although it would be nice if it was all on one line and the `path:line:col` was at the beginning

```
# cargo test |& rg "(error\[\w+\])|(^\s+--> \w+/\w+)" -N --color never
error[E0061]: this function takes 2 parameters but 3 parameters were supplied
   --> src/user/tests.rs:186:37
error[E0061]: this function takes 2 parameters but 3 parameters were supplied
   --> src/export.rs:133:51
error[E0061]: this function takes 2 parameters but 3 parameters were supplied
   --> src/export.rs:139:37
error[E0063]: missing field `revision` in initializer of `export::ArtifactData`
   --> src/export.rs:149:20

```
Thanks!

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #42653.

### CODE REVIEW

**Comment by nrc** [comment]
Rather than adding a new emitter, could you have the human-readable emitter take an argument for detail level - either short or verbose? It seems that we'd save a lot of code dup (and a macro) that way.

Could you put this behind a feature gate?

**Comment by GuillaumeGomez** [comment]
Yes and yes. I was just thinking that we'd want to fully split both of the features but if not...

**Comment by GuillaumeGomez** [comment]
Ok, now just remain to put it behind a feature flag.

**Comment by kennytm** [comment]
Is it possible to make each error only occupy one line, similar to `clang -fno-caret-diagnostics`?

Mock up:

```
$DIR/short-error-format.rs:16:9: error[E0308]: mismatched types
$DIR/short-error-format.rs:18:7: error[E0599]: no method named `salut` found for type `u32` in the current scope
error: aborting due to 2 previous errors 
```

**Comment by eddyb** [comment]
@kenny `mismatched types` is the worst error in the entire compiler, if you can't see the detail lines. I know it looks pretty but it has screwed over pretty much everyone who doesn't look at the console, including the official RLS extension for VS Code.
I [only recently](https://gist.github.com/eddyb/a6f1176f265105105bbecbf4680a3a10) pulled off a hack to have the information available to kate in a manageable way.

**Comment by kennytm** [comment]
@eddyb That sample is referring to the UI test in this PR  https://github.com/GuillaumeGomez/rust/blob/a58becd60d7d7a965ed246cf8e290e2bd94043b9/src/test/ui/short-error-format.stderr. Improving E0308 could happen independent of this PR.

**Comment by eddyb** [comment]
Oh yeah, in that case, I agree, the output in https://github.com/rust-lang/rust/pull/44636#issuecomment-330775555 is a subset of the lines I have from my `errfilt` conversion, but I also have labels/notes (within some limitations), e.g.:
```
src/librustc_trans/mir/lvalue.rs:97:62: error[E0308]: mismatched types
src/librustc_trans/mir/lvalue.rs:97:62: expected struct `rustc::ty::layout::FullLayout`, found enum `rustc::mir::tcx::LvalueTy`
src/librustc_trans/mir/lvalue.rs:97:62: note: expected type `rustc::ty::layout::FullLayout<'_>`
src/librustc_trans/mir/lvalue.rs:97:62: found type `rustc::mir::tcx::LvalueTy<'_>`
```

**Comment by GuillaumeGomez** [comment]
So which output do you want? So I can do it once and for all. :)

**Comment by estebank** [comment]
I feel if the idea is to use as little vertical space the output should be closer to:
```
$DIR/short-error-format.rs:16:9 - error[E0308]: mismatched types
$DIR/short-error-format.rs:18:7 - error[E0599]: no method named `salut` found for type `u32` in the current scope

error: aborting due to 2 previous errors
```

Beyond that, I wonder wether it'd be worth it to try to keep in the diagnostic machinery space for custom single line error messages that would work better for short output (and for text editors). May be I'll just go ahead and introduce it for while and see how much of a pain it becomes. If it is too much, we can rip it out. I do feel strongly _against_ improving things for text editors _in detriment_ of cli output.

**Comment by GuillaumeGomez** [comment]
New output available!

File: src/librustc/session/config.rs [hunk 1]
```diff
@@ -1434,15 +1435,16 @@ pub fn build_session_options_and_crate_config(matches: &getopts::Matches)
...
+            Some("json")  => ErrorOutputType::Json,
+            Some("short") => ErrorOutputType::Short(color),
 
             None => ErrorOutputType::HumanReadable(color),
 
             Some(arg) => {
                 early_error(ErrorOutputType::HumanReadable(color),
-                            &format!("argument for --error-format must be human or json (instead \
-                                      was `{}`)",
+                            &format!("argument for --error-format must be `human` or `json` or \
```
Comment (estebank) [inline src/librustc/session/config.rs]: `human`, `json` or `short`

File: src/librustc/session/mod.rs [hunk 2]
```diff
@@ -833,9 +841,13 @@ pub fn early_error(output: config::ErrorOutputType, msg: &str) -> ! {
     let emitter: Box<Emitter> = match output {
         config::ErrorOutputType::HumanReadable(color_config) => {
             Box::new(EmitterWriter::stderr(color_config,
-                                           None))
+                                           None,
+                                           false))
```
Comment (estebank) [inline src/librustc/session/mod.rs]: Fits in single line?

File: src/librustc/session/mod.rs [hunk 3]
```diff
@@ -846,9 +858,13 @@ pub fn early_warn(output: config::ErrorOutputType, msg: &str) {
...
```
Comment (estebank) [inline src/librustc/session/mod.rs]: Fits in single line?

File: src/librustc/session/mod.rs [hunk 4]
```diff
@@ -676,18 +676,26 @@ pub fn build_session_with_codemap(sopts: config::Options,
...
             Box::new(EmitterWriter::stderr(color_config,
-                                           Some(codemap.clone())))
+                                           Some(codemap.clone()),
+                                           false))
         }
         (config::ErrorOutputType::HumanReadable(_), Some(dst)) => {
             Box::new(EmitterWriter::new(dst,
-                                        Some(codemap.clone())))
+                                        Some(codemap.clone()),
+                                        false))
```
Comment (estebank) [inline src/librustc/session/mod.rs]: Fits in single line?

**Review by estebank — APPROVED** [review]
Some nitpicks, but I'm ok with the PR. Leaving it to @nrc / @eddyb to give final r+.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by alexcrichton** [comment]
@bors: r=estebank

I think @nrc is on leave right now, so @estebank feel free to approve!

**Comment by alexcrichton** [comment]
@bors: r-

er actually, this looks big enough that either @nrc or @eddyb was to review.

Shouldn't this be behind a feature flag?

**Comment by kennytm** [comment]
@alexcrichton You mean `-Z unstable-options`? I don't see how you can put a command line argument behind a feature flag 😅

**Comment by alexcrichton** [comment]
@kennytm indeed!

**Comment by GuillaumeGomez** [comment]
I think @eddyb already said he didn't want to review it so we'll have to wait for @nrc's return or maybe someone from @rust-lang/dev-tools might be able to review it?

**Comment by carols10cents** [comment]
how about....

r? @michaelwoerister

**Comment by michaelwoerister** [comment]
Looks good to me.

@bors r+

**Comment by kennytm** [comment]
`rustfmt` is broken.

```
error[E0061]: this function takes 3 parameters but 2 parameters were supplied
   --> src\tools\rustfmt\src\lib.rs:511:9
    |
511 |         Box::new(Vec::new()),
    |         ^^^^^^^^^^^^^^^^^^^^ expected 3 parameters
   Compiling syntex_errors v0.52.0
error: aborting due to previous error
error: Could not compile `rustfmt-nightly`.
```

**Comment by alexcrichton** [comment]
Triage ping! @GuillaumeGomez

**Comment by GuillaumeGomez** [comment]
The problem is that I need to update `rustfmt` in order to make this PR getting merged but I need this PR to get merged in order to not break `rustfmt`. What's the steps to follow in this case?

**Comment by oli-obk** [comment]
@GuillaumeGomez You need to open a PR against rustfmt (which @nrc needs to place into a branch). Then you can point this PR to the branch on the rustfmt repository. Once this PR is merged, @nrc can merge the branch into rustfmt

**Comment by GuillaumeGomez** [comment]
And done!

**Comment by oli-obk** [comment]
@GuillaumeGomez The rules are specifically that you should be creating a branch on the rustfmt repository. The idea is that if anyone comes back in a year to this commit, they should be able to build the commit successfully. If you delete your repository or even just force push the branch, it won't work anymore.

I wish there was a simpler solution, but it's all we have atm

**Comment by oli-obk** [comment]
Also travis fails:

> patch for `rustfmt-nightly` in `https://github.com/rust-lang/crates.io-index` did not resolve to any crates

which sounds related

**Comment by GuillaumeGomez** [comment]
O.o

I have no idea about what it means. And for my `rustfmt` PR, I'll need someone to create a new branch so I can target it.

**Comment by kennytm** [comment]
What about just break rustfmt in `toolstate.toml`, I think this is now supported.

**Comment by GuillaumeGomez** [comment]
Fine by me!

**Comment by GuillaumeGomez** [comment]
The branch has been created on `rustfmt` so I point to it instead.

**Comment by GuillaumeGomez** [comment]
We now have the following error:

```
[00:00:36] Submodule 'src/tools/rustfmt' (https://github.com/rust-lang-nursery/rustfmt.git) unregistered for path 'src/tools/rustfmt'
[00:00:36] Submodule 'src/tools/rustfmt' (https://github.com/rust-lang-nursery/rustfmt.git) registered for path 'src/tools/rustfmt'
[00:00:37] error: Server does not allow request for unadvertised object fc2c85d423a34086243f019c15e2d9cc2afcbb8f
[00:00:37] Fetched in submodule path 'src/tools/rustfmt', but it did not contain fc2c85d423a34086243f019c15e2d9cc2afcbb8f. Direct fetching of that commit failed.
```

Anyone has an idea?

**Comment by topecongiro** [comment]
@GuillaumeGomez Could you please use `e35e27659dec7a70d1e0e45be2c5f3a02b2c742c`? I think I have erased `fc2c85d423a34086243f019c15e2d9cc2afcbb8f` while I pushed the update :/ I am sorry.

**Comment by GuillaumeGomez** [comment]
Ok, let's hope it'll be the good one this time. :)

**Comment by GuillaumeGomez** [comment]
New error and it doesn't get any better:

```
[00:02:06] extracting /checkout/obj/build/cache/2017-08-29/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
[00:02:06] error: failed to resolve patches for `https://github.com/rust-lang/crates.io-index`
[00:02:06]
[00:02:06] Caused by:
[00:02:06] patch for `rustfmt-nightly` in `https://github.com/rust-lang/crates.io-index` did not resolve to any crates
[00:02:06] failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
[00:02:06] Build completed unsuccessfully in 0:00:46
```

Anyone has an idea?

**Comment by GuillaumeGomez** [comment]
Fixed.

**Comment by alexcrichton** [comment]
@nrc @GuillaumeGomez this landed without a feature flag, was that intentional? I assume not given https://github.com/rust-lang/rust/pull/44636#issuecomment-332858074?

**Comment by nrc** [comment]
I suppose this should be behind `-Zunstable-features` given that comment. I filed https://github.com/rust-lang/rust/issues/45995

Thanks for noticing @alexcrichton !

**Comment by GuillaumeGomez** [comment]
Hum indeed, it should had have a feature flag. I'll add it as soon as possible.

**Comment by vitiral** [comment]
❤️  Thanks for fixing the bug I reported, looking forward to using this when it becomes available!

**Comment by vitiral** [comment]
On [this comment](https://github.com/rust-lang/rust/pull/44636#issuecomment-330798569) my preference would be something like:

```
src/librustc_trans/mir/lvalue.rs:97:62: error[E0308]: mismatched types. Expected struct `rustc::ty::layout::FullLayout`, found enum `rustc::mir::tcx::LvalueTy`
```

It's all in one line but that's probably okay. That's just, like, my opinion though man.

**Pull request merged**